File: subscriber-client/sdk-python/src/stardust_sdk/instrument.py
"""Auto-instrument Anthropic, OpenAI and Google Gen AI clients.

``instrument(client, stardust)`` wraps the client's generation methods in place so
every call is metered, including streaming calls. Wrappers return the vendor's own
response objects unchanged, and a metering failure never breaks the API call.

Covered:

- Anthropic / AsyncAnthropic: ``messages.create`` (with or without ``stream=True``)
- OpenAI / AsyncOpenAI: ``chat.completions.create`` and ``responses.create``.
  Streaming chat completions only report usage if you pass
  ``stream_options={"include_usage": True}``; the SDK doesn't add it for you.
- google-genai ``Client``: ``models.generate_content`` / ``generate_content_stream``
  and the ``client.aio`` equivalents
"""

from __future__ import annotations

import functools
import inspect
import logging
from typing import Any, Callable, Dict, Optional

from .client import Stardust
from .extract import Usage, _get, _int, from_anthropic, from_gemini, from_openai

log = logging.getLogger("stardust_sdk")

_INSTRUMENTED = "__stardust_instrumented__"


# --- stream accumulators ------------------------------------------------------------


class _Accumulator:
    """Collects usage from stream events; ``usage()`` is called once the stream ends."""

    def __init__(self, provider: str, model: Optional[str]):
        self.provider, self.model = provider, model
        self.final: Optional[Usage] = None

    def feed(self, event: Any) -> None:
        raise NotImplementedError

    def usage(self) -> Optional[Usage]:
        return self.final


class _AnthropicAcc(_Accumulator):
    # message_start carries input usage; message_delta carries the running output count.
    def __init__(self, provider: str, model: Optional[str]):
        super().__init__(provider, model)
        self.usage_fields: Dict[str, int] = {}

    def feed(self, event: Any) -> None:
        kind = _get(event, "type")
        if kind == "message_start":
            message = _get(event, "message")
            self.model = _get(message, "model") or self.model
            self._merge(_get(message, "usage"))
        elif kind == "message_delta":
            self._merge(_get(event, "usage"))

    def _merge(self, usage: Any) -> None:
        for name in ("input_tokens", "output_tokens", "cache_read_input_tokens", "cache_creation_input_tokens"):
            value = _get(usage, name)
            if value is not None:
                self.usage_fields[name] = _int(value)

    def usage(self) -> Optional[Usage]:
        if not self.usage_fields:
            return None
        return from_anthropic({"model": self.model, "usage": self.usage_fields}, self.model)


class _OpenAIAcc(_Accumulator):
    # Chat: the last chunk carries usage (with include_usage). Responses: response.completed.
    def feed(self, event: Any) -> None:
        if _get(event, "type") in ("response.completed", "response.incomplete"):
            event = _get(event, "response")
        if _get(event, "usage") is not None:
            self.final = from_openai(event, self.model)


class _GeminiAcc(_Accumulator):
    # Each chunk repeats usage_metadata; the last one holds the totals.
    def feed(self, event: Any) -> None:
        if _get(event, "usage_metadata", "usageMetadata") is not None:
            self.final = from_gemini(event, self.model)


# --- stream proxies -----------------------------------------------------------------


class _StreamProxy:
    """Pass-through wrapper around a vendor stream that records usage when it ends."""

    def __init__(self, stream: Any, acc: _Accumulator, done: Callable[[Optional[Usage]], None]):
        self._stream, self._acc, self._done = stream, acc, done
        self._finished = False

    def _finish(self) -> None:
        if not self._finished:
            self._finished = True
            self._done(self._acc.usage())

    def _feed(self, event: Any) -> None:
        try:
            self._acc.feed(event)
        except Exception:  # noqa: BLE001
            log.exception("Stardust could not read a stream event")

    def __iter__(self):
        try:
            for event in self._stream:
                self._feed(event)
                yield event
        finally:
            self._finish()

    def __enter__(self):
        if hasattr(self._stream, "__enter__"):
            self._stream.__enter__()
        return self

    def __exit__(self, *exc):
        try:
            return self._stream.__exit__(*exc) if hasattr(self._stream, "__exit__") else None
        finally:
            self._finish()

    async def _aiter(self):
        try:
            async for event in self._stream:
                self._feed(event)
                yield event
        finally:
            self._finish()

    def __aiter__(self):
        return self._aiter()

    async def __aenter__(self):
        if hasattr(self._stream, "__aenter__"):
            await self._stream.__aenter__()
        return self

    async def __aexit__(self, *exc):
        try:
            return await self._stream.__aexit__(*exc) if hasattr(self._stream, "__aexit__") else None
        finally:
            self._finish()

    def __getattr__(self, name: str) -> Any:
        return getattr(self._stream, name)


# --- wrapping -------------------------------------------------------------------------


def _wrap(owner: Any, attr: str, stardust: Stardust, provider: str, acc_cls: type,
          extractor: Callable[[Any, Optional[str]], Optional[Usage]], region: Optional[str],
          always_stream: bool = False) -> bool:
    original = getattr(owner, attr, None)
    if original is None or getattr(original, _INSTRUMENTED, False):
        return False

    def record(usage: Optional[Usage]) -> None:
        if usage is not None:
            stardust.record_usage(usage, region=region)

    def handle(result: Any, kwargs: Dict[str, Any]) -> Any:
        model = kwargs.get("model")
        try:
            if always_stream or kwargs.get("stream") is True:
                return _StreamProxy(result, acc_cls(provider, model), record)
            record(extractor(result, model))
        except Exception:  # noqa: BLE001 - metering must never break the caller
            log.exception("Stardust metering failed")
        return result

    # Check the result, not the function: async vendor methods are often plain functions
    # (behind decorators) that return an awaitable.
    @functools.wraps(original)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        result = original(*args, **kwargs)
        if inspect.isawaitable(result):
            async def finish() -> Any:
                return handle(await result, kwargs)
            return finish()
        return handle(result, kwargs)

    setattr(wrapper, _INSTRUMENTED, True)
    setattr(owner, attr, wrapper)
    return True


def _path(obj: Any, *names: str) -> Any:
    for name in names:
        obj = getattr(obj, name, None)
        if obj is None:
            return None
    return obj


def instrument(client: Any, stardust: Stardust, *, region: Optional[str] = None) -> Any:
    """Meter every generation call made through ``client``. Returns the same client.

    ``region`` is the processing region (e.g. ``us-east-1`` for Bedrock); it overrides the
    Stardust client's default for this client only.
    """
    module = type(client).__module__ or ""
    wrapped = 0

    if module.startswith("anthropic"):
        wrapped += _wrap(_path(client, "messages"), "create", stardust, "anthropic", _AnthropicAcc, from_anthropic, region)
    elif module.startswith("openai"):
        wrapped += _wrap(_path(client, "chat", "completions"), "create", stardust, "openai", _OpenAIAcc, from_openai, region)
        wrapped += _wrap(_path(client, "responses"), "create", stardust, "openai", _OpenAIAcc, from_openai, region)
    elif module.startswith("google"):
        for models in (_path(client, "models"), _path(client, "aio", "models")):
            if models is None:
                continue
            wrapped += _wrap(models, "generate_content", stardust, "google", _GeminiAcc, from_gemini, region)
            wrapped += _wrap(models, "generate_content_stream", stardust, "google", _GeminiAcc, from_gemini, region,
                             always_stream=True)
    else:
        raise TypeError(f"Stardust can't instrument {type(client).__name__} from {module!r}")

    if not wrapped:
        log.info("Stardust: nothing new to instrument on %s", type(client).__name__)
    return client

File: subscriber-client/sdk-js/src/instrument.ts
// Auto-instrument Anthropic, OpenAI and Google Gen AI clients.
//
// `instrument(client, stardust)` wraps the client's generation methods in place, so every
// call is metered, including streams. The vendor's own return values come back (with
// promise extras like `.withResponse()` still available), and a metering failure never
// breaks the API call.
//
// - @anthropic-ai/sdk: messages.create (with or without stream: true)
// - openai: chat.completions.create and responses.create. Streaming chat completions only
//   report usage when you pass stream_options: { include_usage: true }.
// - @google/genai: models.generateContent and models.generateContentStream

import type { Stardust } from "./client.ts";
import { fromAnthropic, fromGemini, fromOpenAI, type Provider, type Usage } from "./extract.ts";

type AnyFn = (...args: unknown[]) => unknown;
type Extractor = (r: unknown, model?: string) => Usage | undefined;

const INSTRUMENTED = Symbol.for("stardust.instrumented");

interface Accumulator {
  feed(event: unknown): void;
  usage(): Usage | undefined;
}

function field(obj: unknown, name: string): unknown {
  return obj != null && typeof obj === "object" ? (obj as Record<string, unknown>)[name] : undefined;
}

function anthropicAcc(model?: string): Accumulator {
  // message_start carries input usage; message_delta carries the running output count.
  const usage: Record<string, number> = {};
  let m = model;
  const merge = (u: unknown) => {
    for (const k of ["input_tokens", "output_tokens", "cache_read_input_tokens", "cache_creation_input_tokens"]) {
      const v = field(u, k);
      if (typeof v === "number") usage[k] = v;
    }
  };
  return {
    feed(event) {
      const type = field(event, "type");
      if (type === "message_start") {
        const message = field(event, "message");
        m = (field(message, "model") as string | undefined) ?? m;
        merge(field(message, "usage"));
      } else if (type === "message_delta") {
        merge(field(event, "usage"));
      }
    },
    usage: () => (Object.keys(usage).length ? fromAnthropic({ model: m, usage }, m) : undefined),
  };
}

function lastUsageAcc(extractor: Extractor, model?: string, unwrap?: (e: unknown) => unknown): Accumulator {
  let final: Usage | undefined;
  return {
    feed(event) {
      const e = unwrap ? unwrap(event) : event;
      const u = extractor(e, model);
      if (u) final = u;
    },
    usage: () => final,
  };
}

const openaiUnwrap = (e: unknown) => {
  const type = field(e, "type");
  return type === "response.completed" || type === "response.incomplete" ? field(e, "response") : e;
};

const ACCUMULATORS: Record<Provider, (model?: string) => Accumulator> = {
  anthropic: anthropicAcc,
  openai: (m) => lastUsageAcc((r, mm) => (field(r, "usage") != null ? fromOpenAI(r, mm) : undefined), m, openaiUnwrap),
  google: (m) => lastUsageAcc(fromGemini, m),
};

/** Proxy a vendor async-iterable stream: events pass through, usage is recorded when it ends. */
function proxyStream<T extends object>(stream: T, acc: Accumulator, done: (u: Usage | undefined) => void): T {
  let finished = false;
  const finish = () => {
    if (!finished) {
      finished = true;
      done(acc.usage());
    }
  };
  const source = stream as unknown as AsyncIterable<unknown>;
  async function* iterate() {
    try {
      for await (const event of source) {
        try {
          acc.feed(event);
        } catch (e) {
          console.warn("[stardust] could not read a stream event", e);
        }
        yield event;
      }
    } finally {
      finish();
    }
  }
  return new Proxy(stream, {
    get(target, prop) {
      if (prop === Symbol.asyncIterator) return () => iterate();
      const value = Reflect.get(target, prop, target);
      return typeof value === "function" ? value.bind(target) : value;
    },
  });
}

/** Chain metering onto a vendor promise while keeping its extra methods (e.g. APIPromise.withResponse). */
function keepExtras<T extends object>(original: T, chained: Promise<unknown>): Promise<unknown> {
  return new Proxy(chained, {
    get(target, prop) {
      const source = prop in target ? target : original;
      const value = Reflect.get(source, prop, source);
      return typeof value === "function" ? value.bind(source) : value;
    },
  });
}

function wrap(
  owner: Record<string, unknown> | undefined,
  method: string,
  stardust: Stardust,
  provider: Provider,
  extractor: Extractor,
  region: string | undefined,
  alwaysStream = false,
): boolean {
  const original = owner?.[method] as (AnyFn & { [INSTRUMENTED]?: boolean }) | undefined;
  if (typeof original !== "function" || original[INSTRUMENTED]) return false;

  const record = (u: Usage | undefined) => {
    if (u) stardust.record(u, { region });
  };
  const handle = (result: unknown, params: unknown): unknown => {
    const model = field(params, "model") as string | undefined;
    try {
      if (alwaysStream || field(params, "stream") === true) {
        if (result != null && typeof result === "object" && Symbol.asyncIterator in result) {
          return proxyStream(result, ACCUMULATORS[provider](model), record);
        }
        return result;
      }
      record(extractor(result, model));
    } catch (e) {
      console.warn("[stardust] metering failed", e);
    }
    return result;
  };

  const wrapper = function (this: unknown, ...args: unknown[]) {
    const result = original.apply(this ?? owner, args);
    if (result != null && typeof (result as Promise<unknown>).then === "function") {
      const chained = (result as Promise<unknown>).then((r) => handle(r, args[0]));
      // Metering attaches at call time, so calls read via extras like .withResponse() are metered too.
      // If the call fails and the caller only uses those extras, this branch's rejection must not
      // surface as an unhandled rejection; callers awaiting the result still see the error.
      chained.catch(() => undefined);
      return keepExtras(result as object, chained);
    }
    return handle(result, args[0]);
  } as AnyFn & { [INSTRUMENTED]?: boolean };
  wrapper[INSTRUMENTED] = true;
  owner![method] = wrapper;
  return true;
}

type Client = Record<string, any>; // eslint-disable-line @typescript-eslint/no-explicit-any

/** Meter every generation call made through `client`. Returns the same client. */
export function instrument<T extends object>(client: T, stardust: Stardust, options: { region?: string } = {}): T {
  const c = client as Client;
  const { region } = options;
  if (typeof c.messages?.create === "function" && !c.chat) {
    wrap(c.messages, "create", stardust, "anthropic", fromAnthropic, region);
  } else if (typeof c.chat?.completions?.create === "function") {
    wrap(c.chat.completions, "create", stardust, "openai", fromOpenAI, region);
    wrap(c.responses, "create", stardust, "openai", fromOpenAI, region);
  } else if (typeof c.models?.generateContent === "function") {
    wrap(c.models, "generateContent", stardust, "google", fromGemini, region);
    wrap(c.models, "generateContentStream", stardust, "google", fromGemini, region, true);
  } else {
    throw new TypeError("Stardust can't instrument this client: expected an Anthropic, OpenAI or Google Gen AI client");
  }
  return client;
}
